subprojects/heap-dump/src/main/java/org/gradle/trace/heapdump/HeapDump.java

```diff
@@ -0,0 +1,54 @@
+package org.gradle.trace.heapdump;
+
+import com.sun.management.HotSpotDiagnosticMXBean;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.build.event.BuildEventsListenerRegistry;
+import org.gradle.internal.UncheckedException;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationCompletionListener;
+
+import javax.management.MBeanServer;
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+
+public class HeapDump {
+
+    public HeapDump(GradleInternal gradle, File baseFile) {
+        Provider<HeapDumpService> listenerProvider = gradle.getSharedServices().registerIfAbsent("heap-dump-at-end", HeapDumpService.class, spec -> {
+            spec.getParameters().getBaseFile().set(baseFile.getAbsolutePath());
+        });
+        gradle.getServices().get(BuildEventsListenerRegistry.class).onTaskCompletion(listenerProvider);
+    }
+
+    interface Params extends BuildServiceParameters {
+        Property<String> getBaseFile();
+    }
+
+    public static abstract class HeapDumpService implements OperationCompletionListener, Closeable, BuildService<Params> {
+        static int counter = 0;
+
+        @Override
+        public void onFinish(FinishEvent event) {
+            // Ignore
+        }
+
+        @Override
+        public void close() {
+            try {
+                File dumpFile = new File(getParameters().getBaseFile().get() + "-heap-" + (++counter) + ".hprof");
```

Comment:
Is there any reason to use a build service instead of a `JCmd` invocation? The `jCmd` invocation would also have the advantage that it wouldn't affect the build time to take a heap dump.

Reply:
If we use jcmd (which is what I did initially), then the heap dump happens after all of the build state has been released, which isn't useful for looking at what state is retained during the build. Using a build service allows heap dumps to happen at various points in the build lifecycle. For now, it happens during "build finished" which is good enough for my purposes, but could happen at any number of places.

We can skip heap dumps for warmups via some system property if we want to.